### DISCUSSION TITLE
ExeSQL get tuple index out of range Error

### DISCUSSION THREAD

**Discussion by marcochang1028** [infiniflow/ragflow#2700] [Q&A] [answered]
I am trying to use ExeSQL to perform the sql query, but sometimes it returns the error below.
tuple index out of range Error
From my observation, this error usually happens when the sql statement include LIKE '%XX%'. Here is the simple example which will raise this error

SELECT issue_number, customer_id, complaint_date, issue_description, solution_description FROM customer_issues WHERE solution_description LIKE '%abnormal vibration%'.

I have tried to execute this sql from my python code (psycopg2) and it can return the correct result.

I hope anyone can give me some hint to fix this problem.

**Comment by KevinHuSh** [comment]
I've submit an issue for that.

**Comment by marcochang1028** [comment, marked as answer]
I have confirmed that the problem is caused by the Peewee package, I updated the code in ragflow/agent/component
/exesql.py,
and after switching to pymysql and psycopg2 directly, it now works. 
I've included the code below for your reference. It might not be perfect, but until a proper fix is in place, it can serve as a temporary workaround.

import re
import pandas as pd
import psycopg2
import pymysql
from abc import ABC
from agent.component.base import ComponentBase, ComponentParamBase

class ExeSQLParam(ComponentParamBase):
    """
    Define the ExeSQL component parameters.
    """

    def __init__(self):
        super().__init__()
        self.db_type = "mysql"  # 默認為 MySQL
        self.database = ""
        self.username = ""
        self.host = ""
        self.port = 3306
        self.password = ""
        self.loop = 3
        self.top_n = 30

    def check(self):
        self.check_valid_value(self.db_type, "Choose DB type", ['mysql', 'postgresql', 'mariadb'])
        self.check_empty(self.database, "Database name")
        self.check_empty(self.username, "Database username")
        self.check_empty(self.host, "IP Address")
        self.check_positive_integer(self.port, "IP Port")
        self.check_empty(self.password, "Database password")
        self.check_positive_integer(self.top_n, "Number of records")


class ExeSQL(ComponentBase, ABC):
    component_name = "ExeSQL"

    def _run(self, history, **kwargs):
        if not hasattr(self, "_loop"):
            setattr(self, "_loop", 0)
        if self._loop >= self._param.loop:
            self._loop = 0
            raise Exception("Maximum loop time exceeds. Can't query the correct data via sql statement.")
        self._loop += 1

        # 解析 SQL 語句
        ans = self.get_input()
        ans = "".join(ans["content"]) if "content" in ans else ""
        ans = re.sub(r'^.*?SELECT ', 'SELECT ', repr(ans), flags=re.IGNORECASE)
        ans = re.sub(r';.*?SELECT ', '; SELECT ', ans, flags=re.IGNORECASE)
        ans = re.sub(r';[^;]*$', r';', ans)
        if not ans:
            return ExeSQL.be_output("SQL statement not found!")

        # 根據資料庫類型連接資料庫
        try:
            if self._param.db_type in ["mysql", "mariadb"]:
                conn = pymysql.connect(
                    host=self._param.host,
                    user=self._param.username,
                    password=self._param.password,
                    db=self._param.database,
                    port=self._param.port
                )
            elif self._param.db_type == 'postgresql':
                conn = psycopg2.connect(
                    host=self._param.host,
                    user=self._param.username,
                    password=self._param.password,
                    dbname=self._param.database,
                    port=self._param.port
                )
            cursor = conn.cursor()
            print(f"Connected to {self._param.db_type} database.")
        except Exception as e:
            return ExeSQL.be_output("**Error**: \nDatabase Connection Failed! \n" + str(e))

        # 執行 SQL 語句並處理結果
        sql_res = []
        for single_sql in re.split(r';', ans):
            if not single_sql:
                continue
            try:
                cursor.execute(single_sql)
                rows = cursor.fetchmany(size=self._param.top_n)
                if rows:
                    # 獲取欄位名稱並將結果轉換為 DataFrame
                    columns = [desc[0] for desc in cursor.description]
                    single_res = pd.DataFrame(rows, columns=columns)
                    sql_res.append({"content": "\nTotal: " + str(len(rows)) + "\n" + single_res.to_markdown()})
                else:
                    sql_res.append({"content": "\nNo records found!"})
            except Exception as e:
                sql_res.append({"content": "**Error**:" + str(e) + "\nError SQL Statement:" + single_sql})
        
        conn.close()

        if not sql_res:
            return ExeSQL.be_output("No record in the database!")

        return pd.DataFrame(sql_res)

**Reply by SendDreams** [reply]
Can you help me solve the problem?
RAGFlow Text To SQL Agent Q&a multiple times throw exception:{}

The RAGFlow background reported an error：

2025-02-20 09:14:27,617 INFO 19 172.19.0.6 - - [20/Feb/2025 09:14:27] "POST /v1/canvas/completion HTTP/1.1" 200 -
2025-02-20 09:14:27,893 INFO 19 HTTP Request: POST http://10.16.16.224:11434/api/chat "HTTP/1.1 200 OK"
2025-02-20 09:14:28,558 INFO 19 HTTP Request: POST http://10.16.16.224:11434/api/embeddings "HTTP/1.1 500 Internal Server Error"
2025-02-20 09:14:28,559 ERROR 19 Canvas.run got exception: {}
Traceback (most recent call last):
File "/ragflow/agent/canvas.py", line 223, in prepare2run
ans = cpn.run(self.history, **kwargs)
File "/ragflow/agent/component/base.py", line 420, in run
raise e
File "/ragflow/agent/component/base.py", line 416, in run
res = self._run(history, **kwargs)
File "/ragflow/agent/component/retrieval.py", line 71, in _run
kbinfos = settings.retrievaler.retrieval(query, embd_mdl, kbs[0].tenant_id, self._param.kb_ids,
File "<@beartype(rag.nlp.search.Dealer.retrieval) at 0x7f21f1605e10>", line 35, in retrieval
File "/ragflow/rag/nlp/search.py", line 360, in retrieval
sres = self.search(req, [index_name(tid) for tid in tenant_ids],
File "<@beartype(rag.nlp.search.Dealer.search) at 0x7f21f1605a20>", line 95, in search
File "/ragflow/rag/nlp/search.py", line 108, in search
matchDense = self.get_vector(qst, emb_mdl, topk, req.get("similarity", 0.1))
File "/ragflow/rag/nlp/search.py", line 47, in get_vector
qv, _ = emb_mdl.encode_queries(txt)
File "<@beartype(api.db.services.llm_service.LLMBundle.encode_queries) at 0x7f21eb4916c0>", line 31, in encode_queries
File "/ragflow/api/db/services/llm_service.py", line 246, in encode_queries
emd, used_tokens = self.mdl.encode_queries(query)
File "/ragflow/rag/llm/embedding_model.py", line 269, in encode_queries
res = self.client.embeddings(prompt=text,
File "/ragflow/.venv/lib/python3.10/site-packages/ollama/_client.py", line 201, in embeddings
return self._request(
File "/ragflow/.venv/lib/python3.10/site-packages/ollama/_client.py", line 74, in _request
raise ResponseError(e.response.text, e.response.status_code) from None
ollama._types.ResponseError: {}
2025-02-20 09:14:28,678 INFO 19 172.19.0.6 - - [20/Feb/2025 09:14:28] "GET /v1/canvas/get/1192f206ee7411ef84490242ac130006 HTTP/1.1" 200 -
